### ISSUE TITLE
Has Many relationship Postgres schemas

### ISSUE THREAD

**Issue by ddxtreme** [influitive/apartment#232]
I have some problem which im trying to figure out why this ins't working or if this unsupported.

Im using _**Rails 4.2.1**_

I have 2 models, lets say Account and Projects

I've added the Account to the excluded models

```
class Account < ActiveRecord::Base
  validates :name, :subdomain, presence: true
  validates :subdomain, uniqueness: :true

  has_many :projects, dependent: :destroy

  before_validation :set_subdomain
  after_commit :create_schema


  private
    def set_subdomain
      return true if !self.subdomain.blank? || self.name.blank?
      self.subdomain = self.name.downcase.delete " "
    end

    def create_schema
      Apartment::Tenant.create(self.subdomain)
    end
end

```

And on my Project model:

```
class Project < ActiveRecord::Base
  validates :title, :description, presence: true

  belongs_to :account
end

```

If i try to create a project passing the account_id which is excluded models i get this error:

```
ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  insert or update on table "projects" violates foreign key constraint "fk_rails_cf1be0347a"
DETAIL:  Key (account_id)=(1) is not present in table "accounts".

```

I thought we could have relationships between public models and tenant models.
Any tip? Thanks

**Comment by bradrobertson** [comment]
Is there a reason for the relationship? I mean a `Project` inherently "belongs to" an Account since it's created inside of that tenant. When you're in a specific tenant and call `Project.all`, you'll only get projects that were created in that tenant anyway.

excluded/tenanted relationships work *in theory* but database foreign_key constraints actually remove  that possibility since they would look directly in the current postgresql schema for the association (not the tenanted schema). I'm not suggesting you remove the constraint, more just questioning why the association has to exist.

**Comment by ddxtreme** [comment]
Yes i understand, i went towards that away.
The reason why was mainly about the "default" behaviour we expect, i mean, if i wanted to list all project of the current account, my first thought would be current_account.projects, in this case i have to define the projects method inside the account model to return Project.all.

**Comment by bradrobertson** [comment]
Understood. I think once you get in the mindset of a global current tenant, you won't have to think about nesting your objects under a current_account. It's always implied.

I personally wouldn't even bother with a `.projects` method on `Account` again because of the implied behaviour of all objects already being scoped to the current tenant (Account) that you're on, and just use `Projects.all` directly elsewhere.

Personal preference though.

**Comment by chrsgrrtt** [comment]
This does raise an interesting issue, which we're coming across at the moment, in that you may have 2 global models (Account and User). User may belong to multiple accounts. If the user is then the owner of resources contained within the Account tenancy (say User has many Projects, per Account), then the same error occurs.

If that makes sense?

**Comment by bradrobertson** [comment]
Ya I *think* I understand. You have a User that owns certain projects within an account (tenant) but the foreign keys will be all messed up. Haven't run into that yet as we don't have that in our app. Not sure of the best fix that type of design, other than not using foreign key constraints and coding the queries yourself. Which isn't that much work tbh.

**Comment by chrsgrrtt** [comment]
Yeah, I think for us we'll essentially sacrifice referential integrity on the database, which would come about if as we're adopting uuid's anyway.  

Postgres does support cross-schema references through prefixing I believe, I'll try and grab some time to explore that.

**Comment by nbouscal** [comment]
I've just run into this same issue – we have a model whose data is shared among all tenants, and tenant-specific models need to be able to reference it. Postgres does support cross-schema references, but when you add the FK to the default schema, it won't be a cross-schema reference, so Postgres won't keep that explicit schema reference around when you then copy it over to a new schema. It does work if you run the migrations manually, but not if you use `Apartment::Tenant.create`. I don't know of any way to solve this while still using the `pg_dump` approach to setting up the new schemas. For now I guess I will just accept the loss of referential integrity.

**Comment by betoharres** [comment]
+1 We are using UUID, and the *User* in my project belongs to a *Company*(tenant), and inside that tenant(*Company*) the user has a *Role*, and I can't reference the *User* from inside that tenant without a error 
```Ruby
   (0.2ms)  ROLLBACK
ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  insert or update on table "user_roles" violates foreign key constraint "fk_rails_318345354e"
DETAIL:  Key (user_id)=(e5443277-438a-4ebf-9cf5-46e891a7f2d4) is not present in table "users".
```

**Comment by ronyv89** [comment]
I fixed this issue by removing the reference using a migration and manually adding the account_id through another migration. Everything else remains the same

**Comment by marciogm** [comment]
In your schema.rb
add_foreign_key "projects", "public.accounts", column: :account_id

**Comment by donatoaz** [comment]
Following @marciogm what I did was create a migration. In my case Companies is global and Users are per tenant. I know, I know... not the best idea to have tenanted users table, but I think it will serve me right (I have an admin global model).

```
class FixPkIssueUserCompany < ActiveRecord::Migration
  def self.up
    remove_foreign_key :users, column: :company_id
    add_foreign_key :users, "public.companies", column: :company_id
  end
  
  def self.down
    remove_foreign_key :users, column: :company_id
    add_foreign_key :users, :companies, column: :company_id
  end
end
```

**Comment by Bongs** [comment]
@donatoaz I tried to add foreign key as you mentioned, but it neglects "public." and adds foreign key to table in same schema.

I've following in a migration:
```
create_table :states do |t|
  t.string :name
  t.string :code, limit: 2
end

class CreateParties < ActiveRecord::Migration[5.0]
  def change
    create_table :parties do |t|
      t.string :name
      t.string :type
      t.integer :state_id
      t.timestamps
    end

    add_foreign_key :parties, "public.states", column: :state_id
  end
end
```

But in schema.rb, it says:
`add_foreign_key "parties", "states"`

It doesn't even recognise the column. I've to manually update the schema.

**Comment by cavoixanha** [comment]
I following of @donatoaz. I don't update the schema but It's worked. Because rails migrate to postgres 
This my log:
```
Migrating customer tenant
== 20170714104500 FixFkIssueCategoryToStore: migrating ========================  
-- remove_foreign_key(:categories, {:column=>:store_id})  
   -> 0.0054s  
-- add_foreign_key(:categories, "public.stores", {:column=>:store_id})  
   -> 0.0023s  
== 20170714104500 FixFkIssueCategoryToStore: migrated (0.0080s) ===============
```
I think postges success! If The problem is schema don't generate expected! 
I think don't update schema, if you detected problem notify for me, please!

**Comment by ysyyork** [comment]
Any one used this with rspec? My code can run normally but with rspec it doesn't work. What I did is I created the record which is referred to in public schema inside before(:all) callback and try to set the foreign key in my test case. However it keeps on complaining foreign key constraint violated. I even did pause and open a new console and do it there, it doesn't work. Not sure if it's rspec issue or apartment issue. Anybody has an idea?
